File: dart/lib/src/domain/get_subdomains.dart
import 'dart:typed_data';

import '../constants/addresses.dart';
import '../domain/get_domain_address.dart';
import '../rpc/rpc_client.dart';
import '../utils/deserialize_reverse.dart';
import '../utils/get_reverse_address_from_domain_address.dart';

/// Result containing subdomain and owner information
class SubdomainResult {
  const SubdomainResult({
    required this.subdomain,
    required this.owner,
  });

  /// The subdomain name
  final String subdomain;

  /// The owner address
  final String owner;
}

/// Parameters for getting subdomains
class GetSubdomainsParams {
  const GetSubdomainsParams({
    required this.rpc,
    required this.domain,
  });

  /// The RPC client for blockchain interaction
  final RpcClient rpc;

  /// The domain whose subdomains are to be retrieved
  final String domain;
}

/// Retrieves all subdomains under the specified domain, including their owners.
///
/// This function mirrors js-kit/src/domain/getSubdomains.ts
///
/// [params] - Parameters containing RPC client and parent domain
///
/// Returns a promise that resolves to an array of subdomain objects, each containing the subdomain name and owner address.
Future<List<SubdomainResult>> getSubdomains(GetSubdomainsParams params) async {
  final domainAddressResult = await getDomainAddress(GetDomainAddressParams(
    domain: params.domain,
  ));

  // If the domain is already a subdomain, return empty list
  if (domainAddressResult.isSub) return [];

  final domainAddress = domainAddressResult.domainAddress;

  // Get reverse lookup accounts
  final reversesAsync = params.rpc.getProgramAccounts(
    nameProgramAddress,
    encoding: 'base64',
    filters: [
      MemcmpFilter(
        offset: 0,
        bytes: domainAddress,
        encoding: 'base58',
      ),
      const MemcmpFilter(
        offset: 64,
        bytes: reverseLookupClass,
        encoding: 'base58',
      ),
    ],
  );

  // Get subdomain accounts
  final subsAsync = params.rpc.getProgramAccounts(
    nameProgramAddress,
    encoding: 'base64',
    filters: [
      MemcmpFilter(
        offset: 0,
        bytes: domainAddress,
        encoding: 'base58',
      ),
    ],
    dataSlice: const DataSlice(
      offset: 32,
      length: 32,
    ),
  );

  final results = await Future.wait([reversesAsync, subsAsync]);
  final reverses = results[0];
  final subs = results[1];

  // Create map of reverse addresses to subdomain names
  final reverseMap = <String, String?>{};

  for (final reverse in reverses) {
    final data = reverse.account.data;
    if (data.length >= 96) {
      final reverseData = Uint8List.fromList(data.sublist(96));
      final subdomainName = deserializeReverse(
        reverseData,
        trimFirstNullByte: true,
      );
      reverseMap[reverse.pubkey] = subdomainName;
    }
  }

  // Process subdomains
  final subdomainResults = <SubdomainResult>[];

  for (final sub in subs) {
    try {
      final reverseAddress =
          await getReverseAddressFromDomainAddress(sub.pubkey);
      final subdomainName = reverseMap[reverseAddress];

      if (subdomainName != null) {
        // Extract owner from account data (first 32 bytes)
        final ownerBytes = sub.account.data.take(32).toList();
        final ownerAddress = _base58Encode(ownerBytes);

        subdomainResults.add(SubdomainResult(
          subdomain: subdomainName,
          owner: ownerAddress,
        ));
      }
    } on Exception {
      // Skip invalid subdomains
      continue;
    }
  }

  return subdomainResults;
}

/// Base58 encode helper
String _base58Encode(List<int> input) {
  const alphabet = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

  if (input.isEmpty) return '';

  // Count leading zeros
  var leadingZeros = 0;
  for (var i = 0; i < input.length; i++) {
    if (input[i] == 0) {
      leadingZeros++;
    } else {
      break;
    }
  }

  // Convert to BigInt
  var value = BigInt.zero;
  for (var i = 0; i < input.length; i++) {
    value = value * BigInt.from(256) + BigInt.from(input[i]);
  }

  // Encode to base58
  final result = <String>[];
  final base = BigInt.from(58);

  while (value > BigInt.zero) {
    final remainder = (value % base).toInt();
    result.insert(0, alphabet[remainder]);
    value = value ~/ base;
  }

  // Add leading ones for leading zeros
  for (var i = 0; i < leadingZeros; i++) {
    result.insert(0, '1');
  }

  return result.join();
}

File: js-kit/src/domain/getSubdomains.ts
import {
  Address,
  Base58EncodedBytes,
  GetProgramAccountsApi,
  GetSlotApi,
  Rpc,
} from "@solana/kit";

import { addressCodec, base64Codec } from "../codecs";
import {
  NAME_PROGRAM_ADDRESS,
  REVERSE_LOOKUP_CLASS,
} from "../constants/addresses";
import { assertTldSupported } from "../utils/assertTldSupported";
import { deserializeReverse } from "../utils/deserializers/deserializeReverse";
import { getReverseAddressFromDomainAddress } from "../utils/getReverseAddressFromDomainAddress";
import { getSnsDomainAddress } from "./getSnsDomainAddress";

/**
 * Parameters for retrieving subdomains under a parent domain.
 *
 * @example
 * ```ts
 * const params: GetSubdomainsParams = {
 *   rpc,
 *   domain: "example.sns",
 * };
 * ```
 */
export interface GetSubdomainsParams {
  /** RPC client. */
  rpc: Rpc<GetProgramAccountsApi & GetSlotApi>;
  /** Full parent domain name, including its `.sns` or `.sol` suffix. */
  domain: string;
}

/**
 * A subdomain and the owner recorded in its name registry.
 *
 * @example
 * ```ts
 * const subdomain: GetSubdomainsResult = {
 *   subdomain: "blog",
 *   owner: "Fxuoy3gFjfJALhwkRcuKjRdechcgffUApeYAfMWck6w8" as Address,
 * };
 * ```
 */
export interface GetSubdomainsResult {
  /** TLD-less label recorded by the subdomain's reverse lookup account. */
  subdomain: string;
  /** Owner address stored in the subdomain's name registry account. */
  owner: Address;
}

/**
 * Retrieves subdomains under a parent domain, including their owners.
 *
 * Entries without reverse lookup data are omitted. Passing a subdomain returns
 * an empty array.
 *
 * @param params Subdomain retrieval parameters
 * @param params.rpc RPC client implementing program account lookup
 * @param params.domain Full parent domain name including a `.sns` or `.sol` suffix
 * @returns Subdomain names and owner addresses.
 *
 * @example
 * ```ts
 * const subdomains = await getSubdomains({ rpc, domain: "example.sns" });
 * ```
 */
export const getSubdomains = async ({
  rpc,
  domain,
}: GetSubdomainsParams): Promise<GetSubdomainsResult[]> => {
  const [trimmedDomain] = await assertTldSupported({ rpc, domain });
  const { domainAddress, isSub } = await getSnsDomainAddress({
    domain: trimmedDomain,
  });

  if (isSub) return [];

  const getReversesAsync = rpc
    .getProgramAccounts(NAME_PROGRAM_ADDRESS, {
      encoding: "base64",
      filters: [
        {
          memcmp: {
            offset: 0n,
            bytes: domainAddress as string as Base58EncodedBytes,
            encoding: "base58",
          },
        },
        {
          memcmp: {
            offset: 64n,
            bytes: REVERSE_LOOKUP_CLASS as string as Base58EncodedBytes,
            encoding: "base58",
          },
        },
      ],
    })
    .send();

  const getSubsAsync = rpc
    .getProgramAccounts(NAME_PROGRAM_ADDRESS, {
      encoding: "base64",
      filters: [
        {
          memcmp: {
            offset: 0n,
            bytes: domainAddress as string as Base58EncodedBytes,
            encoding: "base58",
          },
        },
      ],
      dataSlice: { offset: 32, length: 32 },
    })
    .send();

  const [reverses, subs] = await Promise.all([getReversesAsync, getSubsAsync]);

  const map = new Map<string, string | undefined>(
    reverses.map((e) => [
      e.pubkey,
      deserializeReverse({
        data: base64Codec.encode(e.account.data[0]).slice(96),
        trimFirstNullByte: true,
      }),
    ])
  );

  const result = await Promise.all(
    subs.map((sub) =>
      getReverseAddressFromDomainAddress({
        domainAddress: sub.pubkey,
        parentAddress: domainAddress,
      }).then((revKey) => {
        const subdomain = map.get(revKey);
        return subdomain
          ? {
              subdomain,
              owner: addressCodec.decode(
                base64Codec.encode(sub.account.data[0])
              ),
            }
          : undefined;
      })
    )
  );

  const filteredResult = result.filter((sub) => sub !== undefined);

  return filteredResult;
};
